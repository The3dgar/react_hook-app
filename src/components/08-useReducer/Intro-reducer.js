// Estado inicial
const initialState = [
  {
    id: 1,
    todo: "Comprar pan",
    done: false,
  },
];

// Reducer 
const todoReducer = (state = initialState, action) => {
  if (action?.type == "agregar") {
    return [...state, action.payload];
  }
  return state;
};

// Inicializacion
let todos = todoReducer();

// Data para modificar el estado
const newTodo = {
  id: 2,
  todo: "Comprar leche",
  done: false,
};

// Action
const agregarTodoAction = {
  type: "agregar",
  payload: newTodo,
};

// Envio de la accion con la data al reducer
todos = todoReducer(todos, agregarTodoAction);

// Resultado
console.log(todos);
