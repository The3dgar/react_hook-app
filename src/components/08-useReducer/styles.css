body {
  background-color: white;
  color: #333;

}

.list-group-item {
  display: flex;
  align-items: center;
  justify-content: space-between;
}

p{
  cursor: pointer;
  margin: 0px;
}

.complete {
  text-decoration: line-through;
}