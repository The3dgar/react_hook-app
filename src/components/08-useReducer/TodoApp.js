import React, { useReducer, useEffect } from "react";
import "./styles.css";
import { todoReducer } from "./todoReducer";
import { TodoList } from "./TodoList";
import { TodoAdd } from "./TodoAdd";

const init = () => JSON.parse(localStorage.getItem("todos")) || []

export const TodoApp = () => {
  const [todos, dispatch] = useReducer(todoReducer, [], init);

  const handleDelete = (id) => {
    dispatch({ type: "delete", payload: id });
  };

  const handleToggle = (id) => {
    dispatch({ type: "toggle", payload: id });
  };

  const handleAddTodo = (newTodo) => {
    dispatch({ type: "add", payload: newTodo });
  };

  useEffect(() => {
    localStorage.setItem("todos", JSON.stringify(todos));
  }, [todos]);

  return (
    <div>
      <h1>Todo App ({todos.length})</h1>
      <hr></hr>
      <div className="row">
        <div className="col-7">
          <TodoList
            todos={todos}
            handleToggle={handleToggle}
            handleDelete={handleDelete}
          ></TodoList>
        </div>
        <div className="col-5">
          <TodoAdd handleAddTodo ={handleAddTodo}></TodoAdd>
        </div>
      </div>
    </div>
  );
};
