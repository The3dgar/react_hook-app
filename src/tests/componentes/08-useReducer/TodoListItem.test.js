const { shallow } = require("enzyme");
import React from "react";
import "@testing-library/jest-dom";
import { TodoListItem } from "../../../components/08-useReducer/TodoListItem";

import { demoTodos } from "../../fixture/demoTodos";

describe("test in <TodoListItem/>", () => {
  const handleDelete = jest.fn();
  const handleToggle = jest.fn();

  const wrapper = shallow(
    <TodoListItem
      todo={demoTodos[0]}
      index={0}
      handleDelete={handleDelete}
      handleToggle={handleToggle}
    />
  );

  test("should show component", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("should call function handleDelete", () => {
    wrapper.find("button").simulate("click");
    expect(handleDelete).toHaveBeenCalledWith(demoTodos[0].id);
  });

  test("should call function handleToggle", () => {
    wrapper.find("p").simulate("click");
    expect(handleToggle).toHaveBeenCalledWith(demoTodos[0].id);
  });

  test("should show text correctly", () => {
    expect(wrapper.find("p").text().trim().split(" . ")[1]).toBe(demoTodos[0].desc)
  });

  test("should to have complete class", () => {
    const todo = demoTodos[0]
    todo.done = true
    const wrapper = shallow(
      <TodoListItem
        todo={todo}
      />
    );
    expect(wrapper.find("p").hasClass("complete")).toBe(true)
  });
});
