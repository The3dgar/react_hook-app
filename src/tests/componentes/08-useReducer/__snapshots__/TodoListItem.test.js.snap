// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test in <TodoListItem/> should show component 1`] = `
<li
  className="list-group-item"
>
  <p
    className="false"
    onClick={[Function]}
  >
    1
     . 
    Aprender pruebas en react
  </p>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Borrar
  </button>
</li>
`;
